cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,212 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snap-confine-args.h"
+
+#include <string.h>
+
+#include "utils.h"
+
+struct sc_args {
+	// The security tag that the application is intended to run with
+	char *security_tag;
+	// The executable that should be invoked
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc or argv are NULL");
+		goto out;
+	}
+	// Use dereferenced versions of argcp and argvp for convenience.
+	int argc = *argcp;
+	char **const argv = *argvp;
+
+	if (argc == 0) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero");
+		goto out;
+	}
+
+	args = calloc(1, sizeof *args);
+	if (args == NULL) {
+		die("cannot allocate memory for command line arguments object");
+	}
+	// Check if we're being called through the ubuntu-core-launcher symlink.
+	// When this happens we want to skip the first positional argument as it is
+	// the security tag repeated (legacy).
+	bool ignore_first_tag = false;
+	char *basename = strrchr(argv[0], '/');
+	if (basename != NULL) {
+		// NOTE: this is safe because we, at most, may move to the NUL byte
+		// that compares to an empty string.
+		basename += 1;
+		if (strcmp(basename, "ubuntu-core-launcher") == 0) {
+			ignore_first_tag = true;
+		}
+	}
+	// Parse option switches.
+	int optind;
+	for (optind = 1; optind < argc; ++optind) {
+		// Look at all the options switches that start with the minus sign ('-')
+		if (argv[optind][0] != '-') {
+			// On first non-switch argument break the loop. The next loop looks
+			// just for non-option arguments. This ensures that options and
+			// positional arguments cannot be mixed.
+			break;
+		}
+		// Handle option switches
+		if (strcmp(argv[optind], "--version") == 0) {
+			args->is_version_query = true;
+			// NOTE: --version short-circuits the parser to finish
+			goto done;
+		} else if (strcmp(argv[optind], "--classic") == 0) {
+			args->is_classic_confinement = true;
+		} else {
+			// Report unhandled option switches
+			err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+					    "unrecognized command line option: %s",
+					    argv[optind]);
+			goto out;
+		}
+	}
+
+	// Parse positional arguments.
+	//
+	// NOTE: optind is not reset, we just continue from where we left off in
+	// the loop above.
+	for (; optind < argc; ++optind) {
```

Comment:
How is this different from
```C
if (ignore_first_tag) optind++;
args->security_tag = strdup(argv[optind++]);
args->executable = strdup(argv[optind++]);
```
(skipping error checks for clarity). You put yourself in a loop of length two, with an `if (first) {...} else if (second) {...}` as the only body?